feat(op-challenger): Highest Acted L1 Block Number Gauge Metric

refcell opened the pull request:
**Description**

Adds a prometheus gauge metric to the op-challenger to record the highest l1 block number acted upon by the challenger.

**Metadata**

Fixes https://github.com/ethereum-optimism/client-pod/issues/387

refcell committed:
feat(op-challenger): Added l1 block number gauge metric

refcell commented:
Current dependencies on/for this PR:
* `develop`
  * **PR #8724** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8724?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8724?utm_source=stack-comment).

ajsutton commented on op-challenger/game/scheduler/scheduler.go:
```diff
@@ -51,6 +58,7 @@ func NewScheduler(logger log.Logger, m SchedulerMetricer, disk DiskManager, maxC
 		scheduleQueue:  scheduleQueue,
 		jobQueue:       jobQueue,
 		resultQueue:    resultQueue,
+		blockGameMap:   make(map[uint64][]types.GameMetadata),
```
I suspect storing the list of games per block is more than we need.  There's two possible approaches I can see to reduce that:

1. Store the count of pending games for each L1 block. Decrement for each job that completes and then the l1 block is complete when the count gets to 0.
2. Store the highest L1 block completed for each game and then report the lowest value.

The problem is that both of these feel like a memory leak just waiting to happen as we're adding an additional place that we're recording each game and need to do GC on that carefully.

What do you think of trying a version of this where the metrics update is done by the coordinator instead of back in here in the scheduler? The coordinator already maintains a map `states       map[common.Address]*gameState` which tracks the state for each game.  If coordinator.processResult stored the latest process block in that state for each game, it could then set the metric to whatever the minimum l1 block for any game in that map is. Might be easier to test too since the coordinator doesn't mess around with threads and channels.

refcell replied:
This sounds like a good path forward - I agree that the list of games per block is excessive, mostly just done as a quick spike. But the bigger architectural change of moving this into the coordinator seems like a less error-prone approach. Let me continue to spike that out in this pr and then we can see how that looks. Then I'll go ahead and add some tests and cleanup after to move this out of a spike.

ajsutton commented on op-challenger/metrics/metrics.go:
```diff
@@ -195,6 +205,13 @@ func (m *Metrics) RecordGamesStatus(inProgress, defenderWon, challengerWon int)
+	if m.highestActedL1BlockNumber < n {
+		m.highestActedL1BlockNumber = n
+		m.highestActedL1Block.Set(float64(n))
+	}
```
Our metrics are generally thread safe but this won't be.  In this case it wouldn't matter but seems dangerous to have a non-thread safe metric as people won't expect it.  Using an atomic int of some form here is probably the best approach as I do like the fact we can just throw completed L1 block notifications at the metrics in any order and it works it out.

ajsutton reviewed (commented):
Looks like a good start.

refcell commented:
Went ahead and added a few tests since I had some additional time - this is a little different from using the states mapping but still keeps the accounting very minimal using a simple counter.

ajsutton commented on op-challenger/game/scheduler/coordinator.go:
```diff
@@ -90,13 +102,14 @@ func (c *coordinator) schedule(ctx context.Context, games []types.GameMetadata)
 		if err := c.enqueueJob(ctx, j); err != nil {
 			errs = append(errs, fmt.Errorf("failed to enqueue job for game %v: %w", j.addr, err))
 		}
+		c.blockUpdates[j.block]++
```
I think there's a bug here where we only update the latest L1 block if there are jobs to process.  We should be updating it even if we found there were no games requiring action (assuming there aren't any in-flight updates from previous blocks).

ajsutton commented on op-challenger/game/scheduler/coordinator.go:
```diff
@@ -150,6 +163,11 @@ func (c *coordinator) processResult(j job) error {
+	if c.blockUpdates[j.block] == 0 {
+		delete(c.blockUpdates, j.block)
+		c.m.RecordActedL1Block(j.block)
+	}
```
I think this is inaccurate.  If we have a sequence like:

* Block 1 - start processing game A
* Block 2 - start processing game B
* Finishing processing game B <- calls `RecordActedL1Block(2)`
* Finish processing game A <- calls `RecordActedL1Block(1)`

So we're claiming to have processed up to block 2 while game A from block 1 is still in progress.

ajsutton commented on op-challenger/game/scheduler/coordinator.go:
```diff
@@ -33,18 +40,23 @@ type coordinator struct {
+	// blockUpdates maps a block number to the number of jobs that should be scheduled
+	// for that block. This helps us keep track of in-flight jobs and the highest block
+	// that the challenger has fully acted upon (all jobs are complete).
+	blockUpdates map[uint64]uint64
```
hmm, this means we still have a second map to ensure we GC properly, and it needs a different set of GC logic to the states map.  I think maybe you were trying to avoid iterating the states map to find the minimum value to send to metrics, but we will wind up needing to iterate this map to find the minimum anyway and we actually already iterate the states map to remove jobs we no longer require and could just capture the minimum pending job block in that loop (then update based on that and the new jobs we schedule for the current block).

ajsutton commented on op-challenger/game/scheduler/coordinator_test.go:
```diff
@@ -182,6 +181,57 @@ func TestDeleteDataForResolvedGames(t *testing.T) {
+			j.status = types.GameStatusDefenderWon
+		}
+		j.block = 1
+		require.NoError(t, c.processResult(j))
```
The comment says game 3 is still in flight, but the result for it is being returned and the metric is expected to have been updated to indicate that all jobs completed for block 1.

ajsutton commented on op-challenger/game/scheduler/coordinator_test.go:
```diff
@@ -182,6 +181,57 @@ func TestDeleteDataForResolvedGames(t *testing.T) {
+		if j.addr == gameAddr2 {
+			j.status = types.GameStatusDefenderWon
+		}
+		j.block = 1
```
We shouldn't be setting the job block number in any of these tests - it should be set correctly by the scheduler when creating the job.

ajsutton reviewed (commented):
Unless I'm missing something, I don't think this has a couple of bugs.  I took a crack at it in https://github.com/ethereum-optimism/optimism/compare/aj/scheduler-metric-updates.  That latest commit does an extra iteration of the states map which simplifies the logic a fair bit and I think is worth it.  You can do it without introducing an additional iteration though at the cost of more complex logic: https://github.com/ethereum-optimism/optimism/commit/d23a8f2e23dd1ade71b7477ff2bf84bd56079d54

refcell committed:
fix acted block metric

refcell commented:
> Unless I'm missing something, I don't think this has a couple of bugs. I took a crack at it in https://github.com/ethereum-optimism/optimism/compare/aj/scheduler-metric-updates. That latest commit does an extra iteration of the states map which simplifies the logic a fair bit and I think is worth it. You can do it without introducing an additional iteration though at the cost of more complex logic: [d23a8f2](https://github.com/ethereum-optimism/optimism/commit/d23a8f2e23dd1ade71b7477ff2bf84bd56079d54)

This makes sense, I updated it to follow what you did in your commit - it wasn't clear to me if you wanted to merge your work / replace this pr so I just updated the logic here.

ajsutton commented on op-challenger/game/scheduler/coordinator.go:
```diff
@@ -33,24 +41,28 @@ type coordinator struct {
 			return candidate.Proxy == addr
 		}) {
 			delete(c.states, addr)
+			continue
```
nit: We can remove this continue again - at one point I was adding logic into this loop that needed to be skipped but that got removed so the continue is redundant again. Sorry, that's my mess getting left behind. :)

ajsutton commented on op-challenger/metrics/metrics.go:
```diff
@@ -195,6 +207,15 @@ func (m *Metrics) RecordGamesStatus(inProgress, defenderWon, challengerWon int)
 }
 
+func (m *Metrics) RecordActedL1Block(n uint64) {
+	m.bnMutex.Lock()
+	defer m.bnMutex.Unlock()
+	if m.highestActedL1BlockNumber < n {
+		m.highestActedL1BlockNumber = n
+		m.highestActedL1Block.Set(float64(n))
+	}
```
We should be able to remove this logic now and just set the metric.  The coordinator will always be providing the correct value and only increment.

ajsutton reviewed (approved):
LGTM. Just a couple of simplifications.

refcell committed:
challenger cleanup
